jdaviz/configs/specviz/plugins/parsers.py

```diff
@@ -117,28 +121,82 @@ def specviz_spectrum1d_parser(app, data, data_label=None, format=None, show_in_v
                 elif i == 0:
                     app.add_data_to_viewer(spectrum_viewer_reference_name, data_label[i])
 
-        # reset display ranges, or build combined spectrum, when input is a SpectrumList instance
-        if isinstance(data, SpectrumList):
+        if concat_by_file and isinstance(data, SpectrumList):
+            # If >1 spectra in the list were opened from the same FITS file,
+            # group them by their original FITS filenames and add their combined
+            # 1D spectrum to the DataCollection.
+            unique_files = group_spectra_by_filename(app.data_collection)
+            for filename, datasets in unique_files.items():
+                if len(datasets) > 1:
+                    spec = combine_lists_to_1d_spectrum(wlallorig, fnuallorig, dfnuallorig,
+                                                        wave_units, flux_units)
 
-            # build combined spectrum
-            wlallarr = np.array(wlallorig)
-            fnuallarr = np.array(fnuallorig)
-            dfnuallarr = np.array(dfnuallorig)
-            srtind = np.argsort(wlallarr)
-            wlall = wlallarr[srtind]
-            fnuall = fnuallarr[srtind]
-            fnuallerr = dfnuallarr[srtind]
+                    # Make metadata layout conform with other viz.
+                    spec.meta = standardize_metadata(spec.meta)
 
-            # units are not being handled properly yet.
-            unc = StdDevUncertainty(fnuallerr * flux_units)
-            spec = Spectrum1D(flux=fnuall * flux_units, spectral_axis=wlall * wave_units,
-                              uncertainty=unc)
+                    label = f"Combined [{filename}]"
+                    app.add_data(spec, label)
 
-            # Make metadata layout conform with other viz.
-            spec.meta = standardize_metadata(spec.meta)
+                    if show_in_viewer:
+                        app.add_data_to_viewer(spectrum_viewer_reference_name, label)
 
-            # needs perhaps a better way to label the combined spectrum
-            label = "Combined " + data_label[0]
-            app.add_data(spec, label)
-            if show_in_viewer:
-                app.add_data_to_viewer(spectrum_viewer_reference_name, label)
+
+def group_spectra_by_filename(datasets):
+    """
+    Organize the elements of a `~glue.core.data_collection.DataCollection` by the name of
+    the FITS file they came from, if available via the header card "FILENAME".
+
+    Parameters
+    ----------
+    dataset : `~glue.core.data_collection.DataCollection`
+        Collection of datasets
+
+    Returns
+    -------
+    spectra_to_combine : dict
+        Datasets of spectra organized by their filename.
+    """
+    spectra_to_combine = dict()
+
+    for data in datasets:
+        filename = data.meta.get("FILENAME")
+
+        if filename not in spectra_to_combine:
+            spectra_to_combine[filename] = []
```

Comment:
Would `defaultdict(list)` work better or is that not suited for your use case?

Reply:
That's handy, I haven't used that before!